src/java.base/share/classes/java/lang/classfile/TypeKind.java

```diff
@@ -157,4 +157,22 @@ public static TypeKind from(TypeDescriptor.OfField<?> descriptor) {
                 ? fromDescriptor(descriptor.descriptorString())
                 : TypeKind.ReferenceType;
     }
+
+    /**
+     * {@return the type kind associated with the specified class}
+     * @param cl the class
+     */
+    public static TypeKind from(Class<?> cl) {
+        if (cl == null         ) throw new NullPointerException();
+        if (cl == boolean.class) return BooleanType;
+        if (cl == byte.class   ) return ByteType;
+        if (cl == char.class   ) return CharType;
+        if (cl == int.class    ) return IntType;
+        if (cl == long.class   ) return LongType;
+        if (cl == short.class  ) return ShortType;
+        if (cl == float.class  ) return FloatType;
+        if (cl == double.class ) return DoubleType;
+        if (cl == void.class   ) return VoidType;
+        else                     return ReferenceType;
```

Comment:
This can call `Class​::isPrimitive()` to perform an implicit null check and short‑circuit for reference types:
```suggestion
        if (cl.isPrimitive()) { // implicit null check
            if (cl == boolean.class) return BooleanType;
            if (cl == byte.class   ) return ByteType;
            if (cl == char.class   ) return CharType;
            if (cl == int.class    ) return IntType;
            if (cl == long.class   ) return LongType;
            if (cl == short.class  ) return ShortType;
            if (cl == float.class  ) return FloatType;
            if (cl == double.class ) return DoubleType;
            if (cl == void.class   ) return VoidType;
        }
        return ReferenceType;
```

Reply:
https://github.com/openjdk/jdk/pull/20759#issuecomment-2317655533 
Refer to @cl4es 's comment, isPrimitive is a native method and will be slow

Reply:
I said it can be relatively slow in the interpreter. This was an invitation to measure carefully (with and without `-Xint`), not an authoritative statement that we should avoid `Class::isPrimitive` everywhere. I think you need to provide supporting evidence that these kinds of optimizations are worthwhile.